dart-mode.el

```diff
@@ -562,6 +579,251 @@ This uses `dart-format-path' to find the formatter."
       (dart-format-region beg (point)))))
 
 
+;;; Dart analysis server
+
+(defvar dart-debug nil)
+(defvar dart-enable-analysis-server nil)
+(defvar dart-analysis-server nil)
+(defvar dart-analysis-server-snapshot-path
+  "/usr/lib/dart/bin/snapshots/analysis_server.dart.snapshot")
+(defvar dart-analysis-roots nil)
+(defvar dart-analysis-server-next-id 0)
+(defvar dart-analysis-server-callbacks nil)
+
+(defun dart-log (msg)
+  "Logs MSG to the dart log if `dart-debug' is set."
+  (when dart-debug
+    (let* ((log-buffer (get-buffer-create "*dart-debug*"))
+           (iso-format-string "%Y-%m-%dT%T%z")
+           (timestamp-and-log-string
+            (format-time-string iso-format-string (current-time))))
+      (with-current-buffer log-buffer
+        (goto-char (point-max))
+        (insert "\n\n\n")
+        (insert (concat timestamp-and-log-string
+                        "\n"
+                        msg))
+        (insert "\n")))))
+
+(defun dart--start-analysis-server-for-current-buffer ()
+  "Initialize Dart analysis server for current buffer.
+
+This starts Dart analysis server and adds either the pub root
+directory or the current file directory to the analysis roots."
+  (if (not dart-analysis-server) (dart-start-analysis-server))
+  ;; TODO(hterkelsen): Add this file to the priority files.
+  (dart-add-current-file-to-analysis-roots)
+  (dart-add-analysis-server-change-hooks)
+  (add-to-list 'flycheck-checkers 'dart-analysis-server))
+
+(defun dart-start-analysis-server ()
+  "Start the Dart analysis server.
+
+`dart' must be on the PATH."
+  (let ((dart-process
+         (start-process "dart-analysis-server"  "*dart-analysis-server*"
+                        "dart" dart-analysis-server-snapshot-path
+                        "--no-error-notification")))
+    (setq dart-analysis-server (dart-analysis-server-create
+                                dart-process))))
+
+(defun dart-analysis-server-create (process)
+  "Create a Dart analysis server from PROCESS."
+  (let* ((das-buffer (generate-new-buffer (process-name process)))
+         (das-inst (make-dart-analysis-server
```

Comment:
Nit: There's no need for the "das-" names here, since these are local variables. Just use `buffer` and `instance`.